src/engraving/tests/rest_tests.cpp

```diff
@@ -0,0 +1,100 @@
+/*
```

Comment:
I don't entirely agree here. The visual tests are not unit tests, they are in fact system tests. If a vtest fails, it is generally going to take a lot of debugging to find out where the failure comes from; it can be anywhere in the codebase (well, more realistically, anywhere in the engraving module, but still).

This is why unit tests are important _too_. In an ideal world, every method or class is accompanied by a comprehensive collection of unit tests, for just that method or class in isolation. That way, it is immediately clear which parts of the codebase are working correctly and which are not; and many regressions are prevented.

Compared to how we've been working for the past years, that may sound absurdly rigid. But it is the proper way to do it. The fact that we are not doing it is a huge technical debt. The reason that we're not doing it, is, in my perception, that we have never taken the time to clean up all the libmscore-inherited unit tests in the engraving module, and never started adopting a proper testing practice, because the libmscore tests seemed to set such a "bad precedent". (I say "seemed" because I wonder if anyone ever really looked at all of them very closely. There may be sensible stuff there.)

Other advantages of unit tests are that they document how the method should be used, and give a specification of expected results.

However, it should be noted that 100% test coverage is impossible, because UI stuff and stuff that directly uses 3rdparty stuff is hardly unit-testable. According to [an old presentation by Google](https://docs.google.com/presentation/d/1god5fDDd1aP6PwhPodOnAZSPpD80lqYDrHhuhyD7Tvg/htmlpresent#:~:text=Our%20median%20test%20coverage%20is,percentile%20we%20have%2090%25%20coverage.), median test coverage of all their projects is 78%, although their C++ projects are quite a bit below that with 56.6%. 

Another thing to note is that we're not the only software team in the world that neglects unit testing. But that's not really a justification IMO.

Reply:
Something I realized in this project (which I consider to have been well worth the effort for my own uptime in the codebase even though not merged) was that each vtest tests only a single configuration of styles. I wonder if it would be possible to retool them to test more than one. Like, to have rest vtests that test each file with both settings of `multiVoiceRestTwoSpaceOffset`, without having to duplicate all the files to do it.